aslam_cv_pipeline/src/visual-npipeline.cc

```diff
@@ -76,37 +79,45 @@ bool VisualNPipeline::processImageNonBlockingDroppingOldestNFrameIfFull(
     size_t max_output_queue_size) {
   CHECK_GE(max_output_queue_size, 1u);
 
-  bool result = false;
+  if (shutdown_) {
+    return false;
+  }
   std::unique_lock<std::mutex> lock(mutex_);
   if (completed_.size() >= max_output_queue_size) {
     completed_.erase(completed_.begin());
-    result = true;
+    condition_not_full_.notify_all();
   }
-  processImage(camera_index, image, timestamp);
-  return result;
+  processImageImpl(camera_index, image, timestamp);
+  return true;
 }
 
 bool VisualNPipeline::getNextBlocking(std::shared_ptr<VisualNFrame>* nframe) {
   CHECK_NOTNULL(nframe);
-  std::unique_lock<std::mutex> lock(mutex_);
-  if (shutdown_) {
-    return false;
-  }
-  if (completed_.empty()) {
-    condition_not_empty_.wait(lock);
+
+  while (!shutdown_) {
+    std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
again, I believe acquiring the lock needs to go before checking shutdown_. in essence, the processing from checking shutdown_ up until waiting on the condition variable must not be interrupted.